### PR TITLE
allow constructor for density free formular functor

### PR DESCRIPTION
**Pull request by psychocoderHPC**
`FreeFormulaImpl` allows the user functor to define a constructor to move additional information into the profile.
This allow e.g. filtering by cell position relative to the global volume

# example
```C++
//! example implementation is not supporting moving window
struct RelativeRangeProfile
{
    static constexpr float_X start_ratio = 0.1_X;
    static constexpr float_X end_ratio = 0.3_X;
    static constexpr uint32_t usedDim = 1u;

    float_64 first_cell = 0.0_X;
    float_64 last_cell = 0.0_X;
    RelativeRangeProfile()
    {
        const SubGrid< simDim >& subGrid = Environment< simDim >::get().SubGrid();

        auto globalSize = subGrid.getGlobalDomain().size;
        auto deviceOffset = subGrid.getLocalDomain().offset;
        first_cell = ( static_cast< float_64 >( globalSize[usedDim] ) * start_ratio ) -
            static_cast< float_64 >( deviceOffset[ usedDim ] );
        last_cell = ( static_cast< float_64 >( globalSize[ usedDim ] ) * end_ratio ) -
            static_cast< float_64 >( deviceOffset[ usedDim ] );
    }

    //! homogenous density between [start_ratio;end_ratio) of the global domain
    HDINLINE float_X
    operator()(
        floatD_64 const & position_SI,
        float3_64 const & cellSize_SI
    )
    {
        return position_SI[usedDim] >= first_cell && position_SI[usedDim] < last_cell ? 1.0_X : 0.0_X;
    }
};

// definition of free formula profile
using HomogenousRange = FreeFormulaImpl< RelativeRangeProfile >;
```

### CODE REVIEW

**Comment by PrometheusPi** [comment]
Great work - I will test it right away.

**Comment by sbastrakov** [comment]
Looks good to me, not sure if @PrometheusPi has tried this out.

**Comment by PrometheusPi** [comment]
@psychocoderHPC and @sbastrakov sorry I was busy with creating videos and did not test it yet - will do this tomorrow

**Comment by PrometheusPi** [comment]
I am testing it right now.

**Comment by ax3l** [comment]
@PrometheusPi how did the test go? :)

**Comment by PrometheusPi** [comment]
Just an inbetween remark:
Where do these strange velocity fluctuations on the shear surface come from?
grafik

**Comment by PrometheusPi** [comment]
Just to illustrate this a bit further
grafik
Blue is the velocity in one x-direction - red in the other. This strange overlap at the shear surface should not look like this).

This has so far nothing to do with this pull request, but is the current status of the dev to check this pull request.

**Comment by PrometheusPi** [comment]
The issue seen in the isaac visualization, was caused an error on my input set.

**Comment by PrometheusPi** [comment]
Setting density only in (..., [0.25, 0.75], ...) works fine:
grafik

**Comment by PrometheusPi** [comment]
Works like a charm:
grafik

grafik

File: include/picongpu/particles/densityProfiles/FreeFormulaImpl.hpp [hunk 1]
```diff
@@ -21,40 +21,43 @@
...
 #include "picongpu/simulation/control/MovingWindow.hpp"
+#include "picongpu/particles/functor/User.hpp"
 
 
 namespace picongpu
 {
 namespace densityProfiles
 {
-
-template<typename T_ParamClass>
```
[the remark this answers is not in the document]
Comment (sbastrakov) [reply include/picongpu/particles/densityProfiles/FreeFormulaImpl.hpp]: Both before and now the real functor logic is eventually implemented inside the template parameter and then wrapped around by this class. It is still needed, just now called `T_UserFuntor `

Btw it is a typo and should be `T_UserFunctor ` ('c' is missing).

File: include/picongpu/particles/densityProfiles/FreeFormulaImpl.hpp [hunk 2]
```diff
@@ -21,40 +21,43 @@
...
-struct FreeFormulaImpl : public T_ParamClass
-{
-    using ParamClass = T_ParamClass;
-
-    template<typename T_SpeciesType>
-    struct apply
+    template< typename T_UserFuntor >
+    struct FreeFormulaImpl : public particles::functor::User< T_UserFuntor >
```
Comment (sbastrakov) [inline include/picongpu/particles/densityProfiles/FreeFormulaImpl.hpp]: Typo: should be `T_UserFunctor`, not `T_UserFuntor` .

Can we also at least briefly list the requirements for this class?

**Comment by sbastrakov** [comment]
@PrometheusPi on your constructor question - yes, the idea is that with this PR one can pass arbitrary parameters to the user-provided functor through the constructor of `FreeFormulaImpl` class.

**Comment by PrometheusPi** [comment]
@sbastrakov Thank you for the explanation.

**Comment by PrometheusPi** [comment]
@ComputationalRadiationPhysics/picongpu-maintainers after this pull request has been added - should the KHI example be refactored to use 4 instead of 2 species?

**Comment by PrometheusPi** [comment]
@sbastrakov Since @psychocoderHPC is on vacation, I quickly fixed the typo.

**Comment by PrometheusPi** [comment]
*discussed offline with @sbastrakov*:
The documentation of these "template concepts" needs to be done in a wider and more general scope. Thus we will need to postpone this to a future pull request.

**Pull request merged**